File: ormai-ts/src/policy/redaction.ts
/**
 * Field redaction logic.
 *
 * The Redactor applies field-level transformations to query results based on
 * field policies. This includes masking, hashing, and denying field access.
 */

import { createHash } from 'crypto';
import type { FieldPolicy, ModelPolicy } from './models.js';
import { ModelPolicyUtils } from './models.js';

/**
 * Built-in redaction strategies.
 */
export type RedactionStrategy =
  | 'deny'
  | 'mask_email'
  | 'mask_phone'
  | 'mask_card'
  | 'mask_partial'
  | 'hash_sha256';

/**
 * Applies redaction rules to query results.
 *
 * Redaction is applied after query execution to ensure sensitive data
 * never leaves the server in readable form.
 */
export class Redactor {
  constructor(private readonly modelPolicy: ModelPolicy) {}

  /**
   * Apply redaction rules to a single record.
   *
   * Returns a new object with redacted values.
   */
  redactRecord(record: Record<string, unknown>): Record<string, unknown> {
    const result: Record<string, unknown> = {};

    for (const [field, value] of Object.entries(record)) {
      const fieldPolicy = ModelPolicyUtils.getFieldPolicy(this.modelPolicy, field);
      result[field] = this.redactValue(value, fieldPolicy);
    }

    return result;
  }

  /**
   * Apply redaction to a list of records.
   */
  redactRecords(records: Record<string, unknown>[]): Record<string, unknown>[] {
    return records.map((record) => this.redactRecord(record));
  }

  /**
   * Apply redaction to a single value based on policy.
   */
  private redactValue(value: unknown, policy: FieldPolicy): unknown {
    if (value === null || value === undefined) {
      return null;
    }

    switch (policy.action) {
      case 'allow':
        return value;
      case 'deny':
        return null;
      case 'mask':
        return this.applyMask(value, policy.maskPattern);
      case 'hash':
        return this.applyHash(value);
      default:
        return value;
    }
  }

  /**
   * Apply masking to a value.
   */
  private applyMask(value: unknown, pattern: string | undefined): string {
    const strValue = String(value);

    if (pattern) {
      return this.applyCustomMask(strValue, pattern);
    }

    // Default masking based on value format
    if (strValue.includes('@')) {
      return maskEmail(strValue);
    }
    const digitsOnly = strValue.replace(/[+\-\s]/g, '');
    if (/^\d+$/.test(digitsOnly)) {
      if (strValue.length > 10) {
        return maskPhone(strValue);
      }
      return maskPartial(strValue);
    }
    return maskPartial(strValue);
  }

  /**
   * Apply a custom mask pattern.
   */
  private applyCustomMask(value: string, pattern: string): string {
    let result = pattern;

    // Replace {lastN} patterns
    const lastMatch = /\{last(\d+)\}/.exec(pattern);
    if (lastMatch) {
      const n = parseInt(lastMatch[1], 10);
      const lastN = value.length >= n ? value.slice(-n) : value;
      result = result.replace(lastMatch[0], lastN);
    }

    // Replace {firstN} patterns
    const firstMatch = /\{first(\d+)\}/.exec(pattern);
    if (firstMatch) {
      const n = parseInt(firstMatch[1], 10);
      const firstN = value.length >= n ? value.slice(0, n) : value;
      result = result.replace(firstMatch[0], firstN);
    }

    return result;
  }

  /**
   * Hash a value using SHA256.
   */
  private applyHash(value: unknown): string {
    const strValue = String(value);
    return createHash('sha256').update(strValue).digest('hex');
  }
}

/**
 * Mask an email address: user@domain.com -> u***@domain.com
 */
export function maskEmail(email: string): string {
  if (!email.includes('@')) {
    return maskPartial(email);
  }
  const [local, domain] = email.split('@');
  if (local.length <= 1) {
    return `${local}***@${domain}`;
  }
  return `${local[0]}***@${domain}`;
}

/**
 * Mask a phone number: +1234567890 -> +1******890
 */
export function maskPhone(phone: string): string {
  if (phone.length <= 5) {
    return '*'.repeat(phone.length);
  }
  return phone.slice(0, 2) + '*'.repeat(phone.length - 5) + phone.slice(-3);
}

/**
 * Generic partial masking: show first and last char.
 */
export function maskPartial(value: string): string {
  if (value.length <= 2) {
    return '*'.repeat(value.length);
  }
  return value[0] + '*'.repeat(value.length - 2) + value[value.length - 1];
}

/**
 * Mask a credit card number: 1234567890123456 -> ****3456
 */
export function maskCard(card: string): string {
  if (card.length <= 4) {
    return '****';
  }
  return '*'.repeat(card.length - 4) + card.slice(-4);
}

/**
 * Hash a value using SHA256.
 */
export function hashSha256(value: string): string {
  return createHash('sha256').update(value).digest('hex');
}

/**
 * Utility function to mask a value using a built-in strategy.
 *
 * This can be used for custom redaction outside of the main flow.
 */
export function maskValue(value: unknown, strategy: RedactionStrategy): unknown {
  if (value === null || value === undefined) {
    return null;
  }

  const strValue = String(value);

  switch (strategy) {
    case 'deny':
      return null;
    case 'mask_email':
      return maskEmail(strValue);
    case 'mask_phone':
      return maskPhone(strValue);
    case 'mask_card':
      return maskCard(strValue);
    case 'mask_partial':
      return maskPartial(strValue);
    case 'hash_sha256':
      return hashSha256(strValue);
    default:
      return value;
  }
}

/**
 * Create a redactor for a model policy.
 */
export function createRedactor(modelPolicy: ModelPolicy): Redactor {
  return new Redactor(modelPolicy);
}

File: src/ormai/policy/redaction.py
"""
Field redaction logic.

The Redactor applies field-level transformations to query results based on
field policies. This includes masking, hashing, and denying field access.
"""

import hashlib
import re
from enum import Enum
from typing import Any

from ormai.policy.models import FieldAction, FieldPolicy, ModelPolicy


def _mask_partial(value: str) -> str:
    """Generic partial masking: show first and last char."""
    if len(value) <= 2:
        return "*" * len(value)
    return value[0] + "*" * (len(value) - 2) + value[-1]


class RedactionStrategy(str, Enum):
    """Built-in redaction strategies."""

    DENY = "deny"  # Remove field entirely
    MASK_EMAIL = "mask_email"  # user@domain.com -> u***@domain.com
    MASK_PHONE = "mask_phone"  # +1234567890 -> +1******890
    MASK_CARD = "mask_card"  # 1234567890123456 -> ****3456
    MASK_PARTIAL = "mask_partial"  # Show first and last chars
    HASH_SHA256 = "hash_sha256"  # SHA256 hash


def _mask_email_impl(email: str) -> str:
    """Mask an email address: user@domain.com -> u***@domain.com"""
    if "@" not in email:
        return _mask_partial(email)
    local, domain = email.rsplit("@", 1)
    if len(local) <= 1:
        return f"{local}***@{domain}"
    return f"{local[0]}***@{domain}"


def _mask_phone_impl(phone: str) -> str:
    """Mask a phone number: +1234567890 -> +1******890"""
    # Keep first 2 and last 3 chars
    if len(phone) <= 5:
        return "*" * len(phone)
    return phone[:2] + "*" * (len(phone) - 5) + phone[-3:]


class Redactor:
    """
    Applies redaction rules to query results.

    Redaction is applied after query execution to ensure sensitive data
    never leaves the server in readable form.
    """

    def __init__(self, model_policy: ModelPolicy) -> None:
        self.model_policy = model_policy

    def redact_record(self, record: dict[str, Any]) -> dict[str, Any]:
        """
        Apply redaction rules to a single record.

        Returns a new dict with redacted values.
        """
        result = {}
        for field, value in record.items():
            field_policy = self.model_policy.get_field_policy(field)
            result[field] = self._redact_value(field, value, field_policy)
        return result

    def redact_records(self, records: list[dict[str, Any]]) -> list[dict[str, Any]]:
        """Apply redaction to a list of records."""
        return [self.redact_record(record) for record in records]

    def _redact_value(
        self,
        field: str,  # noqa: ARG002
        value: Any,
        policy: FieldPolicy,
    ) -> Any:
        """Apply redaction to a single value based on policy."""
        if value is None:
            return None

        match policy.action:
            case FieldAction.ALLOW:
                return value
            case FieldAction.DENY:
                return None  # Field is removed
            case FieldAction.MASK:
                return self._apply_mask(value, policy.mask_pattern)
            case FieldAction.HASH:
                return self._apply_hash(value)

        return value

    def _apply_mask(self, value: Any, pattern: str | None) -> str:
        """Apply masking to a value."""
        str_value = str(value)

        if pattern:
            return self._apply_custom_mask(str_value, pattern)

        # Default masking based on value format
        if "@" in str_value:
            return _mask_email_impl(str_value)
        if str_value.replace("+", "").replace("-", "").replace(" ", "").isdigit():
            if len(str_value) > 10:
                return _mask_phone_impl(str_value)
            return _mask_partial(str_value)
        return _mask_partial(str_value)

    def _apply_custom_mask(self, value: str, pattern: str) -> str:
        """Apply a custom mask pattern."""
        # Support patterns like "****{last4}" or "{first2}***{last2}"
        result = pattern

        # Replace {last4}, {last2}, etc.
        last_match = re.search(r"\{last(\d+)\}", pattern)
        if last_match:
            n = int(last_match.group(1))
            result = result.replace(last_match.group(0), value[-n:] if len(value) >= n else value)

        # Replace {first4}, {first2}, etc.
        first_match = re.search(r"\{first(\d+)\}", pattern)
        if first_match:
            n = int(first_match.group(1))
            result = result.replace(first_match.group(0), value[:n] if len(value) >= n else value)

        return result

    def _apply_hash(self, value: Any) -> str:
        """Hash a value using SHA256."""
        str_value = str(value)
        return hashlib.sha256(str_value.encode()).hexdigest()


def mask_value(value: Any, strategy: RedactionStrategy) -> Any:
    """
    Utility function to mask a value using a built-in strategy.

    This can be used for custom redaction outside of the main flow.
    """
    if value is None:
        return None

    str_value = str(value)

    match strategy:
        case RedactionStrategy.DENY:
            return None
        case RedactionStrategy.MASK_EMAIL:
            return _mask_email_impl(str_value)
        case RedactionStrategy.MASK_PHONE:
            return _mask_phone_impl(str_value)
        case RedactionStrategy.MASK_CARD:
            return "*" * (len(str_value) - 4) + str_value[-4:] if len(str_value) > 4 else "****"
        case RedactionStrategy.MASK_PARTIAL:
            return _mask_partial(str_value)
        case RedactionStrategy.HASH_SHA256:
            return hashlib.sha256(str_value.encode()).hexdigest()

    return value


def _mask_email(email: str) -> str:
    """Mask an email address: user@domain.com -> u***@domain.com"""
    return _mask_email_impl(email)


def _mask_phone(phone: str) -> str:
    """Mask a phone number: +1234567890 -> +1******890"""
    return _mask_phone_impl(phone)
